frame/im-online/src/lib.rs

```diff
@@ -365,156 +410,130 @@ impl<T: Trait> Module<T> {
 		);
 	}
 
-	pub(crate) fn offchain(now: T::BlockNumber) {
-		let next_gossip = <GossipAt<T>>::get();
-		let check = Self::check_not_yet_gossipped(now, next_gossip);
-		let (curr_worker_status, not_yet_gossipped) = match check {
-			Ok((s, v)) => (s, v),
-			Err(err) => {
-				print(err);
-				return;
-			},
-		};
-		if next_gossip < now && not_yet_gossipped {
-			let value_set = Self::compare_and_set_worker_status(now, false, curr_worker_status);
-			if !value_set {
-				// value could not be set in local storage, since the value was
-				// different from `curr_worker_status`. this indicates that
-				// another worker was running in parallel.
-				return;
-			}
-
-			match Self::do_gossip_at(now) {
-				Ok(_) => {},
-				Err(err) => print(err),
-			}
-		} else {
-			debug::native::debug!(
-				target: "imonline",
-				"Skipping gossip at: {:?} >= {:?} || {:?}",
-				next_gossip,
-				now,
-				if not_yet_gossipped { "not gossipped" } else { "gossipped" }
-			);
+	pub(crate) fn offchain(block_number: T::BlockNumber)
+		-> OffchainResult<T, impl Iterator<Item=OffchainResult<T, ()>>>
+	{
+		let heartbeat_at = <HeartbeatAt<T>>::get();
+		if block_number < heartbeat_at {
+			return Err(OffchainErr::TooEarly(heartbeat_at))
 		}
-	}
 
-	fn do_gossip_at(block_number: T::BlockNumber) -> Result<(), OffchainErr> {
-		// we run only when a local authority key is configured
-		let authorities = Keys::<T>::get();
-		let mut results = Vec::new();
-		let mut local_keys = T::AuthorityId::all();
-		local_keys.sort();
+		Ok(Self::local_authority_keys()
+			.map(move |(authority_index, key)|
+				Self::send_heartbeat(authority_index, key, heartbeat_at, block_number)
+			))
+	}
 
-		for (authority_index, key) in authorities.into_iter()
-			.enumerate()
-			.filter_map(|(index, authority)| {
-				local_keys.binary_search(&authority)
-					.ok()
-					.map(|location| (index as u32, &local_keys[location]))
-			})
-		{
-			if Self::is_online(authority_index) {
-				debug::native::info!(
-					target: "imonline",
-					"[index: {:?}] Skipping sending heartbeat at block: {:?}. Already online.",
-					authority_index,
-					block_number
-				);
-				continue;
-			}
 
+	fn send_heartbeat(
+		authority_index: u32,
+		key: T::AuthorityId,
+		heartbeat_at: T::BlockNumber,
```

Comment:
`heartbeat_at` here seems to be used to identify which session a heartbeat is being sent for. if this is the case, i would just use `session_index` instead which is much clearer and doesn't accidentally encode other information like the delay period before which a session's heartbeat should be sent.